core/src/apps/binance/helpers.py

```diff
@@ -0,0 +1,135 @@
+    elif msg.MESSAGE_WIRE_TYPE == MessageType.BinanceCancelMsg:
+        json_msg = produce_cancel_json(msg)
+    else:
+        raise ValueError("input message unrecognized, is of type " + type(msg).__name__)
+
+    if envelope.source is None or envelope.source <= 0:
+        source = 0
+    else:
+        source = envelope.source
+
+    return ENVELOPE_BLUEPRINT.format(
+        account_number=envelope.account_number,
+        chain_id=envelope.chain_id,
+        memo=envelope.memo,
+        msgs=json_msg,
+        sequence=envelope.sequence,
+        source=source,
+    )
+
+
+def produce_transfer_json(msg: BinanceTransferMsg) -> str:
+    inputs = ""
+    for count, txinput in enumerate(msg.inputs, 1):
+        coins = ""
+        for coincount, coin in enumerate(txinput.coins, 1):
+            coin_json = COIN_BLUEPRINT.format(amount=coin.amount, denom=coin.denom)
+            coins = coins + coin_json
+            if coincount < len(txinput.coins):
+                coins = coins + ","
+        input_json = INPUT_OUTPUT_BLUEPRINT.format(address=txinput.address, coins=coins)
+        inputs = inputs + input_json
+        if count < len(msg.inputs):
+            inputs = inputs + ","
+
+    outputs = ""
+    for count, txoutput in enumerate(msg.outputs, 1):
+        coins = ""
+        for coincount, coin in enumerate(txoutput.coins, 1):
+            coin_json = COIN_BLUEPRINT.format(amount=coin.amount, denom=coin.denom)
+            coins = coins + coin_json
+            if coincount < len(txoutput.coins):
+                coins = coins + ","
+        output_json = INPUT_OUTPUT_BLUEPRINT.format(
+            address=txoutput.address, coins=coins
+        )
+        outputs = outputs + output_json
+        if count < len(msg.outputs):
+            outputs = outputs + ","
+
+    return MSG_TRANSFER_BLUEPRINT.format(inputs=inputs, outputs=outputs)
+
+
+def produce_neworder_json(msg: BinanceOrderMsg) -> str:
+    return MSG_NEWORDER_BLUEPRINT.format(
+        id=msg.id,
+        ordertype=msg.ordertype,
+        price=msg.price,
+        quantity=msg.quantity,
+        sender=msg.sender,
+        side=msg.side,
+        symbol=msg.symbol,
+        timeinforce=msg.timeinforce,
+    )
+
+
+def produce_cancel_json(msg: BinanceCancelMsg) -> str:
+    return MSG_CANCEL_BLUEPRINT.format(
+        refid=msg.refid, sender=msg.sender, symbol=msg.symbol
+    )
+
+
+def address_from_public_key(pubkey: bytes, hrp: str) -> str:
+    """
+    Address = RIPEMD160(SHA256(compressed public key))
+    Address_Bech32 = HRP + '1' + bech32.encode(convert8BitsTo5Bits(RIPEMD160(SHA256(compressed public key))))
+    HRP - bnb for productions, tbnb for tests
+    """
+
+    h = sha256(pubkey).digest()
+    h = ripemd160(h).digest()
+
+    convertedbits = bech32.convertbits(h, 8, 5, False)
+
+    return bech32.bech32_encode(hrp, convertedbits)
+
+
+def validate_full_path(path: list) -> bool:
+    """
+    Validates derivation path to equal 44'/714'/a'/0/0,
```

Comment:
Do we need the 5-part path? Is that what Ledger uses?
Otherwise it would be better to use Stellar-like 3-part path

Reply:
it looks like ledger checks all 5 parts if I understand the code correctly
https://github.com/binance-chain/ledger-app-binance/blob/8781c371f16ebd750250bc38841e232e7c67fcfe/src/app_main.c#L165

Reply:
ok, please add an entry to docs/coins/README